Add a helper function to create tensorboard crd

jingzhang36 opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/pipelines/1237)
<!-- Reviewable:end -->

Ark-kun commented on frontend/server/k8s-helper.ts:
```diff
@@ -127,3 +131,30 @@ export function getPodLogs(podName: string): Promise<string> {
+    kind: 'Viewer',
+    metadata: {
+      generateName: 'viewer-',
+      namespace: 'kubeflow',
```
I do not think we should hard-code the namespace.

jingzhang36 replied:
When I used one-click deployment of kubeflow, I found all the resulting pods in "kubeflow" namespace. E.g., ml-pipeline-ui pod, ml-pipeline pod.  I guess the namespace is configurable via some yaml when deployment is performed? If that is the case, I can follow the same convention here if feasible.

Ark-kun replied:
@IronPan How should we get the namespace properly?

IronPan replied:
talked offline with @jingzhang36 for more context. it should not be hardcoded. 
Here contains the example how to customize the namespace
https://github.com/kubeflow/pipelines/tree/master/manifests

jingzhang36 commented:
Related issue https://github.com/kubeflow/pipelines/issues/1100

rileyjbauer commented on frontend/server/k8s-helper.ts:
```diff
@@ -127,3 +131,30 @@ export function getPodLogs(podName: string): Promise<string> {
+  const namespace = 'kubeflow';
+  const plural = 'tensorboard';
+  const body = {
+    apiVersion: 'kubeflow.org/v1beta1',
```
Any reason not to use the above constants here? (and below)

jingzhang36 replied:
A quick question: if I pass "group" and "version" parameters to the method createNamespacedCustomObject, do I still need to specify "apiVersion" field in "body" parameter? Or I can just leave "apiVersion" field unset in "body"?

jingzhang36 replied:
Done with using the constants

rileyjbauer commented on frontend/server/k8s-helper.ts:
```diff
@@ -127,3 +131,30 @@ export function getPodLogs(podName: string): Promise<string> {
+      }
+    }
+  };
+  const pretty = null;
```
`pretty` is an optional parameter to the `createNamespacedCustomObject`, so you can just ignore it and call it with
`createNamespacedCustomObject(group, version, namespace, plural, body)`

jingzhang36 replied:
Done

jingzhang36 commented:
Updates: 
(1) createTensorboardInstance, getTensorboardInstance and waitforTensorboardInstance are counterparts for createTensorboardPod, getTensorboardAddress and  waitForTensorboard.
(2). getTensorboardHandler and createTensorboardHandler in server.ts are switching to use create/get/waitfor TensorboardInstance.

IronPan commented on frontend/server/k8s-helper.ts:
```diff
@@ -127,3 +132,79 @@ export function getPodLogs(podName: string): Promise<string> {
+  const pods = (await k8sV1CustomObjectClient.listNamespacedCustomObject(group,
+    version, namespace, plural)).body.items;
+  const pod = pods.find((p) =>
+    p.metadata.name == 'viewer-' + crypto.SHA1(logdir) &&
```
will crypto.SHA1(logdir) have size limit?
the metadata name have size limit of 256 as far as i know. it might fail if the length of the logdir is too long

jingzhang36 replied:
My understanding is that the output length of sha1 is fixed at 160 bit as long as the input message is < 2^64 bits. https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf

IronPan commented on frontend/server/k8s-helper.ts:
```diff
@@ -127,3 +132,79 @@ export function getPodLogs(podName: string): Promise<string> {
+    apiVersion: group + '/' + version,
+    kind: 'Viewer',
+    metadata: {
+      name: 'viewer-' + crypto.SHA1(logdir),
```
can you extract the 'viewer-' + crypto.SHA1(logdir) into a function? and reuse the function below

jingzhang36 replied:
done

IronPan commented on frontend/server/k8s-helper.ts:
```diff
@@ -127,3 +132,79 @@ export function getPodLogs(podName: string): Promise<string> {
+
+  // TODO: take the configuration below to a separate file
+  // Name of the viewer resource is based on logDir.
+  const group = 'kubeflow.org';
```
these const are defined twice in this file. consider share them in this file

jingzhang36 replied:
done

IronPan commented on frontend/server/k8s-helper.ts:
```diff
@@ -127,3 +132,79 @@ export function getPodLogs(podName: string): Promise<string> {
+  const group = 'kubeflow.org';
+  const version = 'v1beta1';
+  const plural = 'viewers';
+  const pods = (await k8sV1CustomObjectClient.listNamespacedCustomObject(group,
```
I think we should look for crd than pod. pod might not start yet or not scheduled due to resource limits but crd should always exist.

jingzhang36 replied:
changed to "viewers" and "viewer", the name of our custom resources.

IronPan commented:
/lgtm

IronPan commented:
/retest

IronPan commented:
/retest

jingzhang36 commented:
/retest

jingzhang36 commented:
/assign @IronPan I updated the PR with a version that passes end-to-end test, please take another look
/assign @neuromage I changed the viewer service's port from 80 to 6006, please review
/assign @rileyjbauer I made a change to the encoded uri in tensorboard.tsx, and some comment explaining my change, please take a look

IronPan commented on frontend/server/k8s-helper.ts:
```diff
@@ -124,6 +134,102 @@ export function getPodLogs(podName: string): Promise<string> {
+    throw new Error('Cannot access kubernetes Custom Object API');
+  }
+
+  const viewers = (await k8sV1CustomObjectClient.listNamespacedCustomObject(
```
does client lib has getNamespacedCustomObject?
if so can we use that instead?

jingzhang36 replied:
done

jingzhang36 commented on frontend/server/k8s-helper.ts:
```diff
@@ -124,6 +134,102 @@ export function getPodLogs(podName: string): Promise<string> {
+    return;
+  }
+
+  // TODO: take the configuration below to a separate file
```
since the removal of credentials, the body is short and seems no need to put it to some place separate. so remove the TODO

jingzhang36 commented on frontend/server/k8s-helper.ts:
```diff
@@ -124,6 +134,102 @@ export function getPodLogs(podName: string): Promise<string> {
+        spec: {
+          containers: [{
+            env: [{
+              name: 'GOOGLE_APPLICATION_CREDENTIALS',
```
done

jingzhang36 commented on frontend/server/k8s-helper.ts:
```diff
@@ -124,6 +134,102 @@ export function getPodLogs(podName: string): Promise<string> {
+    v.metadata.name == getNameOfViewerResource(logdir) &&
+    v.spec.tensorboardSpec.logDir == logdir &&
+    v.spec.type == 'tensorboard');
+  return viewer ? `http://${viewer.metadata.name}-service.kubeflow.svc.cluster.local:6006/tensorboard/${viewer.metadata.name}/` : '';
```
done

jingzhang36 commented:
/assign @ironpan

jingzhang36 commented:
/test kubeflow-pipeline-sample-test

IronPan commented:
/lgtm
/approve

IronPan commented:
This change looks good. 
One thing need to call out - Once this is released, some customers can't use the latest version since this CRD requires a ClusterRole. We should prioritize to ensure the CRD can be run in a specific namespace.

jingzhang36 commented:
> This change looks good.
> One thing need to call out - Once this is released, some customers can't use the latest version since this CRD requires a ClusterRole. We should prioritize to ensure the CRD can be run in a specific namespace.

In CRD, the namespace field is parametrized (at const body { metadata { namespace ... ). Is that sufficient to ensure CRD is run on some namespace, which is derived from context?

IronPan commented:
> In CRD, the namespace field is parametrized (at const body { metadata { namespace ... ). Is that sufficient to ensure CRD is run on some namespace, which is derived from context?

That's the field to specify what viewer instance to deploy to. It's not serve for the purpose I was mentioning. The goal is for controller to be able to be run within a specific namespace.

IronPan commented:
/test kubeflow-pipeline-sample-test

jingzhang36 commented:
> > In CRD, the namespace field is parametrized (at const body { metadata { namespace ... ). Is that sufficient to ensure CRD is run on some namespace, which is derived from context?
> 
> That's the field to specify what viewer instance to deploy to. It's not serve for the purpose I was mentioning. The goal I mentioned is for controller to be able to be run within a specific namespace.

I see. So the controller (viewer controller) needs to be namespaced too, e.g., in the same namespace as the rest pods such as ml-pipeline, ml-pipeline-ui etc?
